### ISSUE TITLE
Please support reporting BatchItemFailures when processing streams

### ISSUE THREAD

**Issue by gheibia** [aws/aws-lambda-dotnet#903]
## The Question

Lambda supports optimizing batch processing with custom checkpoints. That is, for example, when a function is trigged by a DynamoDB stream, one could return a "sequenceNumber" in the batch and have Lambda split the batch at that point and retry (instead of recursively splitting the batch in the middle).
Summarizing this article (https://aws.amazon.com/blogs/compute/optimizing-batch-processing-with-custom-checkpoints-in-aws-lambda/), the function will have to return a JSON response containing a single item called "batchItemFailures".

This SDK doesn't seem to support that. If it does, I haven't been able to find anything about this in the documentation.

Also, if it's not supported in the SDK currently, I'm wondering if there is a workaround for this.

### Environment

* Targeted .NET Platform:  netcoreapp3.1

---

This is a :question: general question

**Comment by ashishdhingra** [comment]
Hi @gheibia,

Good afternoon.

Thanks for posting guidance question. Looking at the article [Optimizing batch processing with custom checkpoints in AWS Lambda](https://aws.amazon.com/blogs/compute/optimizing-batch-processing-with-custom-checkpoints-in-aws-lambda/), the only thing this Lambda high level library doesn't appear to support is the model for returning the object containing a single item `batchItemFailures` which should be translated into following JSON (for example):
```JSON
{ 
  "batchItemFailures": [ 
    {
      "itemIdentifier": sequenceNumber
    }
  ]
}
```

The details of DynamoDB event is as per the documentation at [Using AWS Lambda with Amazon DynamoDB](https://docs.aws.amazon.com/lambda/latest/dg/with-ddb.html) (also mentioned in the code comment in [DynamoDBEvent](https://github.com/aws/aws-lambda-dotnet/blob/master/Libraries/src/Amazon.Lambda.DynamoDBEvents/DynamoDBEvent.cs)). The documentation also lists your scenario in section **Reporting batch item failures**. 

Taking some guidance from the Java example, you could model your class for reporting batch item failures to something like below:
```C#
namespace Amazon.Lambda.DynamoDBEvents
{
    using System;
    using System.Collections.Generic;
    using System.Runtime.Serialization;

    /// <summary>
    /// This class is used as the return for AWS Lambda functions that are invoked by DynamoDB to report batch item failures.
    /// https://docs.aws.amazon.com/lambda/latest/dg/with-ddb.html#services-ddb-batchfailurereporting
    /// </summary>
    [DataContract]
    public class StreamsEventResponse
    {
        [DataMember(Name = "batchItemFailures", EmitDefaultValue = false)]
#if NETCOREAPP3_1
        [System.Text.Json.Serialization.JsonPropertyName("batchItemFailures")]
#endif
        public IList<BatchItemFailure> BatchItemFailures { get; set; }

        /// <summary>
        /// The class representing the BatchItemFailure.
        /// </summary>
        [DataContract]
        public class BatchItemFailure
        {
            /// <summary>
            /// Sequence number of the record which failed processing.
            /// </summary>
            [DataMember(Name = "itemIdentifier", EmitDefaultValue = false)]
#if NETCOREAPP3_1
            [System.Text.Json.Serialization.JsonPropertyName("itemIdentifier")]
#endif
            public string ItemIdentifier { get; set; }
        }
    }
}
```
and then modify your Lambda function signature to below (taking example from https://github.com/aws/aws-lambda-dotnet/tree/master/Libraries/src/Amazon.Lambda.DynamoDBEvents):
```C#
public class Function
{
    public StreamsEventResponse Handler(DynamoDBEvent ddbEvent)
    {
        var batchItemFailures = new List<StreamsEventResponse.BatchItemFailure>();
        string curRecordSequenceNumber = "";
  
        foreach (var record in ddbEvent.Records)
        {
            try
            {
                //Process your record
                var ddbRecord = record.Dynamodb;
                var keys = string.Join(", ", ddbRecord.Keys.Keys);
                curRecordSequenceNumber = ddbRecord.SequenceNumber;
                Console.WriteLine($"{record.EventID} - Keys = [{keys}], Size = {ddbRecord.SizeBytes} bytes");
            }
            catch(Exception e)
            {
                //Return failed record's sequence number
                batchItemFailures.add(new StreamsEventResponse.BatchItemFailure(curRecordSequenceNumber));
                return new StreamsEventResponse(batchItemFailures);
            }
        }
        Console.WriteLine($"Successfully processed {ddbEvent.Records.Count} records.");
        return new StreamsEventResponse(batchItemFailures);
    }
}
```

Please let me know if it works for you. If yes, I think it makes sense to submit the PR to add support for this model object in this repository.

Thanks,
Ashish

**Comment by gheibia** [comment]
Thanks. I've tested the code and it seems to be working. Producing an error in my environment to test the splitting case proved to be more difficult. I'll update this ticket once I have more update.

**Comment by ashishdhingra** [comment]
> Thanks. I've tested the code and it seems to be working. Producing an error in my environment to test the splitting case proved to be more difficult. I'll update this ticket once I have more update.

@gheibia We are in the process of introducing this model definition for batch item failure response in `Amazon.Lambda.DynamoDBEvents` package. I will update here once it is done so that you would not have to use separate hand written class for this scenario.

**Comment by jasonmc86** [comment]
@ashishdhingra how about for the same functionality for connecting lambda with Kinesis Streams?

**Comment by ashishdhingra** [comment]
> @ashishdhingra how about for the same functionality for connecting lambda with Kinesis Streams?

@jasonmc86 It's a good ask. Would you like contribute and submit a PR which could be reviewed by the team? In case it is not feasible, feel free to open a new issue with `feature-request` label mentioning all the details so that it could be tracked and prioritized.
